site/content/blog/2021-04-23-optimizing-go-compiler-2.md

```diff
@@ -0,0 +1,343 @@
+
+As we learned in the first part, the issue we need to investigate is that the generated code for the following function is not as optimized as it could be:
+
+```go
+import "encoding/binary"
+
+func copyArrayToSlice(dest []byte, src *[8]byte) {
+	_ = dest[8]                                // 1. bounds check
+	temp := binary.LittleEndian.Uint64(src[:]) // 2. read the contents of the source array into temp
+	binary.LittleEndian.PutUint64(dest, temp)  // 3. write the contents into the destiny slice
+}
+```
+
+Let's use our first tool, SSA, to understand what the compiler is generating. If we generate the SSA for the original function, `copyArrayToSlice`, and look at the `lower` phase, we'll see the following:
+
+```
+b1:
+
+	v1 (?) = InitMem <mem>
+	v7 (5) = Arg <*[8]byte> {src} (b.ptr[*byte], src[*[8]byte])
+	v8 (?) = MOVQconst <int> [8] (b.cap[int], b.len[int])
+	v9 (+6) = Arg <int> {dest} [8] (b.len[int], dest+8[int])
+	v248 (84) = Arg <*byte> {dest}
+	v153 (6) = CMPQconst <flags> [8] v9
+UGT v153 → b43 b3 (likely) (6)
+
+b3: ← b1
+
+	v12 (6) = LoweredPanicBoundsC <mem> [0] v8 v9 v1
+Exit v12 (6)
+
+b24:
+BlockInvalid (+83)
+
+b25:
+BlockInvalid (84)
+
+b27:
+BlockInvalid (85)
+
+b29:
+BlockInvalid (86)
+
+b31:
+BlockInvalid (87)
+
+b33:
+BlockInvalid (88)
+
+b35:
+BlockInvalid (89)
+
+b37:
+BlockInvalid (90)
+
+b39:
+BlockInvalid (91)
+
+b43: ← b1
+
+	v18 (+7) = LoweredNilCheck <void> v7 v1
+	v29 (7) = InlMark <void> [0] v1
+	v152 (+8) = InlMark <void> [1] v1
+	v34 (+79) = MOVQload <uint64> v7 v1
+	v171 (+84) = MOVBstore <mem> v248 v34 v1
+	v168 (+85) = SHRQconst <uint64> [8] v34
+	v216 (+89) = SHRQconst <uint64> [40] v34
+	v180 (+91) = SHRQconst <uint64> [56] v34
+	v219 (88) = MOVLstore <mem> [1] v248 v168 v171
+	v243 (90) = MOVWstore <mem> [5] v248 v216 v219
+	v255 (91) = MOVBstore <mem> [7] v248 v180 v243
+Ret v255 (+8)
+```
+
+The interesting blocks are `b1`, which initialize the values with the arguments of the function, and `b43`, which makes the actual copy.
+
+## Definition block
+
+Let's take a closer look at the block where the values are initialized, `b1`:
+
+```
+v1 (?) = InitMem <mem>
+v7 (5) = Arg <*[8]byte> {src} (b.ptr[*byte], src[*[8]byte])
+v8 (?) = MOVQconst <int> [8] (b.cap[int], b.len[int])
+v9 (+6) = Arg <int> {dest} [8] (b.len[int], dest+8[int])
+v248 (84) = Arg <*byte> {dest}
+v153 (6) = CMPQconst <flags> [8] v9
+```
+
+There are a couple of formalities there, such as representing the initialized memory with the value `v1`, but then we see how we define the arguments of the function:
+
+```
+v7 (5) = Arg <*[8]byte> {src} (b.ptr[*byte], src[*[8]byte])
+```
+
+`v7` is the second argument of the function, `src`, which is a pointer to an array of 8 bytes (remember that `src` was of type `[8]byte`).
+
+```
+v248 (84) = Arg <*byte> {dest}
+```
+
+`v248` is the first argument of the function, `dest`, which is a pointer to bytes (we do not know the length here, as it was defined as a slice, which has a variable length).
```

Comment:
It is not really a "pointer to bytes", it is a "slice header" - which is comprised of three things - the base address of the array, length, and capacity. That's what those three lines - v8, v9, v248 are I think. Anyways, I think it's perhaps simpler to just skip this section.

Reply:
I was going to remove the section per your previous comment, but then read this one and realized I had not understood this block completely. So, instead of removing the section, I now think I should actually expand it :blush:

But I need some help with this. Could you let me know if this makes sense:

```
v1 (?) = InitMem <mem>
```

Init memory, easy peasy.

```
v7 (5) = Arg <*[8]byte> {src} (b.ptr[*byte], src[*[8]byte])
```

This represents the `src` array. The `<*[8]byte> is the type. I *guess* `{src}` is the GCnode. But I'm not sure what `(b.ptr[*byte], src[*[8]byte])` is, actually. I don't even know what that `b` is.

```
v8 (?) = MOVQconst <int> [8] (b.cap[int], b.len[int])
```

I guess this has to do with the length of the slice. Are we simply storing an `8` into the slice capacity and length?

```
v9 (+6) = Arg <int> {dest} [8] (b.len[int], dest+8[int])
```

Again, not sure what this argument is. We have three `Arg` lines, but the function only gets two arguments. I think this has to do with the bounds check in the first line of the function, because `v9` is then used in the `CMPQconst` line to decide which block we're going to next, but I'm not sure what the exact meaning of this line is.

```
v248 (84) = Arg <*byte> {dest}
```

This is a pointer to where the *data* of the `dest` slice lives, right?

```
v153 (6) = CMPQconst <flags> [8] v9
```

Finally, we're comparing `v9` to 8 (?) Is `v9` the length of the slice, maybe?

Reply:
I can probably guess some parts of it, but I am nowhere near an SME on this matter. Let me ping @josharian to lend some of his expertise on this.